### ISSUE TITLE
Server closes the connection with a FIN packet immediately after sending a GO_AWAY frame. All accepted streams upto the lastStreamId are not getting processed

### ISSUE THREAD

**Issue by Sumi264** [reactor/reactor-netty#2735]
My application server using http2 reactive netty server (version: 1.0.26) closes the connection immediately after sending a GO_AWAY packet, without processing outstanding requests up to the lastStreamId. 

+--- org.springframework.boot:spring-boot-starter-webflux:2.7.7
|    +--- org.springframework.boot:spring-boot-starter:2.7.7 (*)
|    +--- org.springframework.boot:spring-boot-starter-json:2.7.7 (*)
|    +--- org.springframework.boot:spring-boot-starter-reactor-netty:2.7.7
|    |    \--- io.projectreactor.netty:reactor-netty-http:1.0.26

[https://www.rfc-editor.org/rfc/rfc7540#section-6.8](RFC-7540) states that all "Activity on streams numbered lower or equal to the last stream identifier might still complete successfully.  The sender of a GOAWAY frame might gracefully shut down a connection by sending a GOAWAY frame, maintaining the connection in an "open" state until all in-progress streams complete."

Does reactive netty server support processing of all accepted streams upto the lastStreamId sent in GO_AWAY frame before terminating the connection? If yes, then please advise if some configuration is needed to make reactive netty http2 server to ensure the processing of all streams up to the lastStreamId quoted in GO_AWAY frame.

**Comment by violetagg** [comment]
@Sumi264 Can you describe your scenario in more details? Is this a graceful shutdown of the server or some other connection close scenario?

**Comment by Sumi264** [comment]
Yes, graceful shutdown is configured for the application server. The application is running in a containerized environment in a kubenetes pod. I'm deleting the pod to observe the behaviour of server with graceful shutdown. 


        ```
factory.addServerCustomizers(httpServer -> httpServer.idleTimeout(Duration.ofMillis(idleTimeout)));
        factory.addServerCustomizers(builder -> builder.protocol(HttpProtocol.H2C, HttpProtocol.HTTP11));
        factory.setShutdown(Shutdown.GRACEFUL);
```

**Comment by Sumi264** [comment]
@violetagg  I'm pasting the logs snippet where it's visible that GO_AWAY frame has lastStreamId=2127, and INBOUND DATA is seen for this streamId, but no OUTBOUND DATA. The connection gets closed as soon as the GO_AWAY signal is received, without processing this lastStream.


`"reactor.netty.http.server.h2","message":"[id: 0xb7b87797, L:/192.168.219.77:8080 - R:/10.75.213.193:21174] INBOUND DATA: streamId=2123 padding=0 endStream=false length=1648 bytes=7b0a20202020226e66496e7374616e63654964223a202263623539633438632d613266612d313165392d613261332d326132616532646263636534222c0a2020..."
"reactor.netty.http.server.h2","message":"[id: 0xb7b87797, L:/192.168.219.77:8080 - R:/10.75.213.193:21174] INBOUND DATA: streamId=2123 padding=0 endStream=true length=0 bytes="
"reactor.netty.http.server.h2","message":"[id: 0xb7b87797, L:/192.168.219.77:8080 - R:/10.75.213.193:21174] OUTBOUND DATA: streamId=2123 padding=0 endStream=true length=18 bytes=7b44656661756c7420526573706f6e73657d"
"reactor.netty.http.server.h2","message":"[id: 0xb7b87797, L:/192.168.219.77:8080 - R:/10.75.213.193:21174] INBOUND DATA: streamId=2125 padding=0 endStream=false length=1648 bytes=7b0a20202020226e66496e7374616e63654964223a202263623539633438632d613266612d313165392d613261332d326132616532646263636534222c0a2020..."
"reactor.netty.http.server.h2","message":"[id: 0xb7b87797, L:/192.168.219.77:8080 - R:/10.75.213.193:21174] INBOUND DATA: streamId=2125 padding=0 endStream=true length=0 bytes="
"reactor.netty.http.server.h2","message":"[id: 0xb7b87797, L:/192.168.219.77:8080 - R:/10.75.213.193:21174] OUTBOUND DATA: streamId=2125 padding=0 endStream=true length=18 bytes=7b44656661756c7420526573706f6e73657d"
"reactor.netty.http.server.h2","message":"[id: 0xb7b87797, L:/192.168.219.77:8080 - R:/10.75.213.193:21174] INBOUND DATA: streamId=2127 padding=0 endStream=false length=1648 bytes=7b0a20202020226e66496e7374616e63654964223a202263623539633438632d613266612d313165392d613261332d326132616532646263636534222c0a2020..."
"reactor.netty.http.server.h2","message":"[id: 0xb7b87797, L:/192.168.219.77:8080 - R:/10.75.213.193:21174] INBOUND DATA: streamId=2127 padding=0 endStream=true length=0 bytes="
"reactor.netty.http.server.h2","message":"[id: 0xb7b87797, L:/192
[wk_gracefulshutdown_inout_filtered.txt](https://github.com/reactor/reactor-netty/files/11025700/wk_gracefulshutdown_inout_filtered.txt)
.168.219.77:8080 - R:/10.75.213.193:21174] OUTBOUND GO_AWAY: lastStreamId=2127 errorCode=0 length=0 bytes="
"reactor.netty.http.server.HttpServer","message":"[b7b87797, L:/192.168.219.77:8080 - R:/10.75.213.193:21174] CLOSE"
"reactor.netty.http.server.HttpServer","message":"[b7b87797, L:/192.168.219.77:8080 ! R:/10.75.213.193:21174] INACTIVE"
"reactor.netty.http.server.HttpServer","message":"[b7b87797, L:/192.168.219.77:8080 ! R:/10.75.213.193:21174] UNREGISTERED"`

**Comment by Sumi264** [comment]
Hi @violetagg ,
Thanks for the fix.
I was trying to test it in my environment, but seems the issue is not resolved. 

I was going through the fix that was provided for this issue, and came across some comments saying '//"FutureReturnValueIgnored" this is deliberate'.

I want to give a brief description of my application server (built on reactive netty using spring-webflux), which exposes a rest api with the below signature.

```
@RequestMapping("/api/requests")
public CompletableFuture<ResponseEntity<?>> handleRequest(RequestEntity<byte[]> reqEntity,
			@RequestHeader HttpHeaders reqHeaders) {...}

```
This API is responsible for receiving the requests and handing them over to an executor thread pool, which does some processing.
Once the business logic processing is done by the executor threads, the CompletableFuture is marked as complete with ResponseEntity. 

In the cases where CompletableFuture is used in the RestController, will the fix provided work? 
Please advise.

**Comment by violetagg** [comment]
@Sumi264 Please either provide some reproducible example or at least some logs where `wiretap` is enabled.

**Comment by Sumi264** [comment]
@violetagg  Please find the logs with wiretap enabled. 
[netty-streams-not-processed-goaway.log](https://github.com/reactor/reactor-netty/files/11059022/netty-streams-not-processed-goaway.log)

Here, 
OUTBOUND GO_AWAY: lastStreamId=12837 , but there's only inbound data logs for this streamId, no outbound data log is observed.

**Comment by Sumi264** [comment]
@violetagg Sure, let me try it out, and share the logs with you. Thanks!!

**Comment by violetagg** [comment]
@Sumi264 Any chance for a reproducible example?

**Comment by Sumi264** [comment]
@violetagg I'll try to share a reproducible example soon.

**Comment by violetagg** [comment]
@Sumi264 You are right we need to send `GO_AWAY` to notify the client to stop opening streams and then we need to keep the connection "open" until we process all streams up to the last stream id sent with `GO_AWAY`. Thanks a lot for the reproducible example.
Will you be able to try this PR #2758

**Comment by Sumi264** [comment]
Hi @violetagg ,
I have done initial testing and it seems to be working.
I'm further planning to test with High traffic.
I've a query as below 
-> Is there a time limit till which Netty will wait for the lastStreamId to be processed? Or the wait is purely based on last StreamId processed?

Thanks!!

**Comment by violetagg** [comment]
> Hi @violetagg , I have done initial testing and it seems to be working. I'm further planning to test with High traffic.

That's very good. Thanks!

> I've a query as below -> Is there a time limit till which Netty will wait for the lastStreamId to be processed? Or the wait is purely based on last StreamId processed?

The timeout that is used for `disposeNow` is the limit. Once reaching the timeout there will be a force close.

**Comment by Sumi264** [comment]
Hi @violetagg  
Is the timeout value being passed in disposeNow(timeout) API for graceful termination configurable? 
It would be helpful if it's configurable since, in different application this graceful termination period can be configured as per the application environment's need.

I'm running this application in a k8s pod, built with springboot(as shown in the earlier attached reproducible example).
Will it be possible to provide an API in "WebServerFactoryCustomizer<NettyReactiveWebServerFactory>" to set this graceful termination period?
For example this class has an API to set the shutdwon to graceful, as shown below.
factory.setShutdown(Shutdown.GRACEFUL);

It would be nice to have any api like above where user can pass the graceful termination period in seconds.

**Comment by violetagg** [comment]
@Sumi264 For questions related to Spring you should contact Spring maintainers. Quick search shows that there is such configuration already https://docs.spring.io/spring-boot/docs/current/reference/htmlsingle/#web.graceful-shutdown

**Comment by Sumi264** [comment]
@violetagg 
My question is, whether the timeout value being passed in **disposeNow(timeout)** API user configurable?
What is the timeout value passed in disposeNow(timeout) api?

I'm using "spring.lifecycle.timeout-per-shutdown-phase=20s" already in my application, does it mean, disposeNow(timeout) api will consider timeout as 20s?

**Comment by violetagg** [comment]
I don't know the integration in Spring Boot.
https://github.com/spring-projects/spring-boot/search?q=disposeNow
Here it seems they call `disposeNow` with `Duration.ofNanos(Long.MAX_VALUE)`

**Comment by Sumi264** [comment]
Hi @violetagg 

While testing the above fix with high traffic I observed some connections are closing without a GoAway.
Sometimes the connection is closed with FIN, also sometimes a connection is closed by RST or sometimes with FIN followed by RST.
I'm attaching the logs for your reference to help me understand the reason behind this behavior.

Test 1: 20 Connections and 200 MPS Traffic
1 connection was closed by FIN without a GO_AWAY attaching logs for reference.
log file: direct_FIN_reference.log
[direct_FIN_reference.log](https://github.com/reactor/reactor-netty/files/11260982/direct_FIN_reference.log)

Test 2: 100 Connections and 1000 MPS Traffic with 150ms sleep delay in netty application.
Observed connections closed by RST and FIN without GO_AWAY attaching logs for reference.
log file: RST_close.log, FIN_close.log
[RST_close.log](https://github.com/reactor/reactor-netty/files/11260981/RST_close.log)
[FIN_close.log](https://github.com/reactor/reactor-netty/files/11260987/FIN_close.log)

**Comment by pderop** [comment]
Hi @Sumi264 ,

I don't see any CLOSE in the log files. Regarding the RST or FIN, is it sent by the client or by the server ?

Can you tell how to reproduce this ? is it with your [reproducer nettyyest](https://github.com/reactor/reactor-netty/files/11115045/nettytest.zip) ? should I run a load using the "netty/test/goaway" URI or using "netty/streams" ?

thanks

**Comment by Sumi264** [comment]
Hi @pderop,
I'm attaching the tcpdump for the connection getting terminated without GO_AWAY, with either an RST or FIN, where it is visible that RST, or FIN is sent by server.
[reactor-netty-RST_FIN.zip](https://github.com/reactor/reactor-netty/files/11269540/reactor-netty-RST_FIN.zip)
It contains below three files.
1. FIN_CLOSE.pcap, RST_Close.pcap
Server(sample application server built with reactive netty embedded in spring-webflux) => 192.168.8.14, Port:8080
Client => 10.75.212.241, Port:52279
2. direct_FIN_reference.pcap
Server: 10.75.213.193, Port:8080
Client: 10.75.213.193, Port:52646
3. RST_Close.pcap
Server(sample application server built with reactive netty embedded in spring-webflux) => 192.168.8.14, Port:8080
Client => 10.75.212.241, Port:56548

For reproducing this issue, [NettyServer-sample-app](https://github.com/reactor/reactor-netty/files/11115045/nettytest.zip) can be used, and load can be run on "netty/streams" API. 

Please let me know if any further info is needed.

**Comment by pderop** [comment]
thanks a lot @Sumi264;

FYI, I'm investigating since yesterday, will try to get back to you soon.

**Comment by pderop** [comment]
I have reproduced, thanks @Sumi264;

We might have a problem when the server is configured with H2C, and HTTP1.1 
I'm working on a candidate PR, will probably create it tomorrow.
thanks.

**Comment by pderop** [comment]
I have attempted to do a PR (see #2777); will you be able to test it ? 
I can't merge this PR this week, because it must be discussed with the team next week; but that would be nice if you could rebuild it and confirm if it works ?

thanks @Sumi264 !

**Comment by Sumi264** [comment]
Sure @pderop, let me try it out. Thanks !!

**Comment by Sumi264** [comment]
Hi @pderop, I tried testing it with 2 MPS and 4 MPS traffic. It seems to be working fine now.
Thanks !!

**Comment by Sumi264** [comment]
Hi @violetagg and @pderop,
Could you please tell me in which release of reactor-netty will the complete fix for this issue be available?

Thanks in advance!

**Comment by violetagg** [comment]
@Sumi264 The release is planned for tomorrow.

**Comment by move02** [comment]
Hi @violetagg .

I had similar problem with my server configured only HTTP11. 
The server was using reactor 1.1.6 with spring-boot-starter-webflux 3.0.6.

I'm testing graceful shutdown to apply on my production server but while testing, sometimes last request seems to received before the server started to graceful shutdown phase but the channel has disconnected right after received it.

`[2023-06-29 09:45:15.489] DEBUG o.s.c.s.DefaultLifecycleProcessor - Stopping beans in phase 2147482623
[2023-06-29 09:45:15.490] INFO o.s.b.w.e.netty.GracefulShutdown - Commencing graceful shutdown. Waiting for active requests to complete
[2023-06-29 09:45:15.491] DEBUG r.netty.transport.ServerTransport - [e70d9d27, L:/[0:0:0:0:0:0:0:0%0]:8182] Server is about to be disposed with timeout: PT2562047H47M16.854775807S
[2023-06-29 09:45:15.495] DEBUG r.n.http.server.HttpServerOperations - [ea2ac38d, L:/10.42.105.58:8182 - R:/10.42.32.192:6412] Increasing pending responses, now 1
[2023-06-29 09:45:15.495] DEBUG r.netty.http.server.HttpServer - [ea2ac38d-7, L:/10.42.105.58:8182 - R:/10.42.32.192:6412] Handler is being applied: org.springframework.http.server.reactive.ReactorHttpHandlerAdapter@539ccecf
[2023-06-29 09:45:15.496] DEBUG o.s.w.s.a.HttpWebHandlerAdapter - [ea2ac38d-2174] HTTP GET "/foo?num=2168"
[2023-06-29 09:45:15.496] DEBUG o.s.w.r.r.m.a.RequestMappingHandlerMapping - [ea2ac38d-2174] Mapped to org.test.Controller#foo(int)
[2023-06-29 09:45:15.496] DEBUG o.s.w.r.r.m.a.ResponseBodyResultHandler - [ea2ac38d-2174] Using 'text/plain;charset=UTF-8' given [*/*] and supported [text/plain;charset=UTF-8, text/event-stream, text/plain;charset=UTF-8, */*]
[2023-06-29 09:45:15.496] DEBUG o.s.w.r.r.m.a.ResponseBodyResultHandler - [ea2ac38d-2174] 0..1 [java.lang.String]
[2023-06-29 09:45:15.496] INFO org.ncsoft.Controller - received: 2168
[2023-06-29 09:45:15.497] DEBUG o.s.core.codec.CharSequenceEncoder - [3e3040cb-1845] Writing "bar - 1840"
[2023-06-29 09:45:15.497] DEBUG r.n.http.server.HttpServerOperations - [3e3040cb-6, L:/10.42.105.58:8182 - R:/10.42.32.192:19472] Decreasing pending responses, now 0
[2023-06-29 09:45:15.497] DEBUG o.s.w.s.a.HttpWebHandlerAdapter - [3e3040cb-1845] Completed 200 OK
[2023-06-29 09:45:15.497] DEBUG r.n.http.server.HttpServerOperations - [3e3040cb-6, L:/10.42.105.58:8182 - R:/10.42.32.192:19472] Last HTTP response frame
[2023-06-29 09:45:15.497] DEBUG r.netty.channel.ChannelOperations - [3e3040cb-6, L:/10.42.105.58:8182 - R:/10.42.32.192:19472] [HttpServer] Channel inbound receiver cancelled (subscription disposed).
[2023-06-29 09:45:15.497] DEBUG r.n.http.server.HttpServerOperations - [3e3040cb-6, L:/10.42.105.58:8182 - R:/10.42.32.192:19472] Last HTTP packet was sent, terminating the channel
**[2023-06-29 09:45:15.497] DEBUG r.netty.channel.ChannelOperations - [ea2ac38d-7, L:/10.42.105.58:8182 ! R:/10.42.32.192:6412] [HttpServer] Channel inbound receiver cancelled (channel disconnected).**

// ... processing other requests

[2023-06-29 09:45:20.483] DEBUG o.s.core.codec.CharSequenceEncoder - [cc7bb073-2173] Writing "bar - 2167"
[2023-06-29 09:45:20.483] DEBUG r.n.http.server.HttpServerOperations - [cc7bb073-7, L:/10.42.105.58:8182 - R:/10.42.32.192:26041] Decreasing pending responses, now 0
[2023-06-29 09:45:20.483] DEBUG o.s.w.s.a.HttpWebHandlerAdapter - [cc7bb073-2173] Completed 200 OK
[2023-06-29 09:45:20.483] DEBUG r.n.http.server.HttpServerOperations - [cc7bb073-7, L:/10.42.105.58:8182 - R:/10.42.32.192:26041] Last HTTP response frame
[2023-06-29 09:45:20.483] DEBUG r.netty.channel.ChannelOperations - [cc7bb073-7, L:/10.42.105.58:8182 - R:/10.42.32.192:26041] [HttpServer] Channel inbound receiver cancelled (subscription disposed).
[2023-06-29 09:45:20.483] DEBUG r.n.http.server.HttpServerOperations - [cc7bb073-7, L:/10.42.105.58:8182 - R:/10.42.32.192:26041] Last HTTP packet was sent, terminating the channel
[2023-06-29 09:45:20.483] INFO o.s.b.w.e.netty.GracefulShutdown - Graceful shutdown complete
[2023-06-29 09:45:20.483] DEBUG o.s.c.s.DefaultLifecycleProcessor - Stopping beans in phase 2147481599
[2023-06-29 09:45:20.484] DEBUG o.s.c.s.DefaultLifecycleProcessor - Bean 'webServerGracefulShutdown' completed its stop procedure
[2023-06-29 09:45:20.484] DEBUG r.netty.transport.ServerTransport - [e70d9d27, L:/[0:0:0:0:0:0:0:0%0]:8182] Server has been disposed
[2023-06-29 09:45:20.484] DEBUG o.s.c.s.DefaultLifecycleProcessor - Bean 'webServerStartStop' completed its stop procedure
[2023-06-29 09:45:20.484] DEBUG o.s.c.s.DefaultLifecycleProcessor - Stopping beans in phase -2147483647
[2023-06-29 09:45:20.484] DEBUG o.s.c.s.DefaultLifecycleProcessor - Bean 'springBootLoggingLifecycle' completed its stop procedure
[2023-06-29 09:45:20.484] DEBUG o.s.s.c.ThreadPoolTaskExecutor - Shutting down ExecutorService 'applicationTaskExecutor'
`

After read this issue, i updated the server's reactor netty version to 1.1.8 and last request are handled normally within graceful shutdown phase.

[this pr](https://github.com/reactor/reactor-netty/pull/2777) includes fixes when the server only configured in HTTP11 too?

**Comment by violetagg** [comment]
@move02 The mentioned PR is related to HTTP/2 only.

**Comment by move02** [comment]
Sorry for deleting my comment. The comment was this PR also related when HTTP11 server also and it wasn't.

Thanks for replying.
